<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jaewon Kim">

<title>class14</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="class14_files/libs/clipboard/clipboard.min.js"></script>
<script src="class14_files/libs/quarto-html/quarto.js"></script>
<script src="class14_files/libs/quarto-html/popper.min.js"></script>
<script src="class14_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="class14_files/libs/quarto-html/anchor.min.js"></script>
<link href="class14_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="class14_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="class14_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="class14_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="class14_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">class14</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jaewon Kim </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#|eval: FALSE</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
다음의 패키지를 부착합니다: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
다음의 패키지를 부착합니다: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
다음의 패키지를 부착합니다: 'IRanges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:grDevices':

    windows</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: SummarizedExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: MatrixGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
다음의 패키지를 부착합니다: 'MatrixGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>필요한 패키지를 로딩중입니다: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
다음의 패키지를 부착합니다: 'Biobase'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>metafile <span class="ot">&lt;-</span> <span class="st">"GSE37704_metadata.csv"</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>countfile <span class="ot">&lt;-</span> <span class="st">"GSE37704_featurecounts.csv"</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>colData <span class="ot">=</span> <span class="fu">read.csv</span>(metafile, <span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>countData <span class="ot">=</span> <span class="fu">read.csv</span>(countfile, <span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(colData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              condition
SRR493366 control_sirna
SRR493367 control_sirna
SRR493368 control_sirna
SRR493369      hoxa1_kd
SRR493370      hoxa1_kd
SRR493371      hoxa1_kd</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(countData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                length SRR493366 SRR493367 SRR493368 SRR493369 SRR493370
ENSG00000186092    918         0         0         0         0         0
ENSG00000279928    718         0         0         0         0         0
ENSG00000279457   1982        23        28        29        29        28
ENSG00000278566    939         0         0         0         0         0
ENSG00000273547    939         0         0         0         0         0
ENSG00000187634   3214       124       123       205       207       212
                SRR493371
ENSG00000186092         0
ENSG00000279928         0
ENSG00000279457        46
ENSG00000278566         0
ENSG00000273547         0
ENSG00000187634       258</code></pre>
</div>
</div>
<p>Q. Complete the code below to remove the troublesome first column from countData</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>countData <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(countData[, <span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(countData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                SRR493366 SRR493367 SRR493368 SRR493369 SRR493370 SRR493371
ENSG00000186092         0         0         0         0         0         0
ENSG00000279928         0         0         0         0         0         0
ENSG00000279457        23        28        29        29        28        46
ENSG00000278566         0         0         0         0         0         0
ENSG00000273547         0         0         0         0         0         0
ENSG00000187634       124       123       205       207       212       258</code></pre>
</div>
</div>
<p>Q. Complete the code below to filter countData to exclude genes (i.e.&nbsp;rows) where we have 0 read count across all samples (i.e.&nbsp;columns).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>zero.ind <span class="ot">&lt;-</span> <span class="fu">rowSums</span>((countData[ ,<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(countData[<span class="dv">1</span>,]) ] <span class="sc">==</span> <span class="dv">0</span>)) <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>countData <span class="ot">&lt;-</span> countData[<span class="sc">!</span>zero.ind, ]</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(countData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                SRR493366 SRR493367 SRR493368 SRR493369 SRR493370 SRR493371
ENSG00000279457        23        28        29        29        28        46
ENSG00000187634       124       123       205       207       212       258
ENSG00000188976      1637      1831      2383      1226      1326      1504
ENSG00000187961       120       153       180       236       255       357
ENSG00000187583        24        48        65        44        48        64
ENSG00000187642         4         9        16        14        16        16</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">=</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData=</span> countData,</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">colData=</span> colData,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">design=</span> <span class="sc">~</span>condition)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in DESeqDataSet(se, design = design, ignoreRank): some variables in
design formula are characters, converting to factors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">=</span> <span class="fu">DESeq</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 13282 6 
metadata(1): version
assays(4): counts mu H cooks
rownames(13282): ENSG00000279457 ENSG00000187634 ... ENSG00000276345
  ENSG00000271254
rowData names(22): baseMean baseVar ... deviance maxCooks
colnames(6): SRR493366 SRR493367 ... SRR493370 SRR493371
colData names(2): condition sizeFactor</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span> <span class="fu">results</span>(dds, <span class="at">contrast=</span><span class="fu">c</span>(<span class="st">"condition"</span>, <span class="st">"hoxa1_kd"</span>, <span class="st">"control_sirna"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Q. Call the summary() function on your results to get a sense of how many genes are up or down-regulated at the default 0.1 p-value cutoff.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 13282 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4333, 33%
LFC &lt; 0 (down)     : 4400, 33%
outliers [1]       : 0, 0%
low counts [2]     : 0, 0%
(mean count &lt; 1)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( res<span class="sc">$</span>log2FoldChange, <span class="sc">-</span><span class="fu">log</span>(res<span class="sc">$</span>padj) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class14_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Q. Improve this plot by completing the below code, which adds color and axis labels</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a color vector for all genes</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>mycols <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"gray"</span>, <span class="fu">nrow</span>(res) )</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Color red the genes with absolute fold change above 2</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>standard <span class="ot">&lt;-</span> <span class="fu">abs</span>(res<span class="sc">$</span>log2FoldChange) <span class="sc">&gt;</span> <span class="dv">2</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>mycols[standard] <span class="ot">&lt;-</span> <span class="st">"red"</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Color blue those with adjusted p-value less than 0.01</span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a><span class="co">#  and absolute fold change more than 2</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>inds <span class="ot">&lt;-</span> (res<span class="sc">$</span>padj <span class="sc">&lt;</span> <span class="fl">0.01</span>) <span class="sc">&amp;</span> (standard)</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>mycols[inds] <span class="ot">&lt;-</span> <span class="st">"blue"</span></span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( res<span class="sc">$</span>log2FoldChange, <span class="sc">-</span><span class="fu">log</span>(res<span class="sc">$</span>padj), <span class="at">col=</span> mycols, <span class="at">xlab=</span><span class="st">"Log2(FoldChange)"</span>, <span class="at">ylab=</span><span class="st">"-Log(P-value)"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class14_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Q. Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results by completing the code below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"AnnotationDbi"</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"org.Hs.eg.db"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">columns</span>(org.Hs.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"         
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
[26] "UNIPROT"     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>symbol <span class="ot">=</span> <span class="fu">mapIds</span>(org.Hs.eg.db,</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keys=</span> <span class="fu">row.names</span>(res),</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keytype=</span><span class="st">"ENSEMBL"</span>,</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">column=</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">multiVals=</span><span class="st">"first"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>entrez <span class="ot">=</span> <span class="fu">mapIds</span>(org.Hs.eg.db,</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keys=</span> <span class="fu">row.names</span>(res),</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">keytype=</span><span class="st">"ENSEMBL"</span>,</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">column=</span> <span class="st">"ENTREZID"</span>,</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">multiVals=</span><span class="st">"first"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>genename <span class="ot">=</span> <span class="fu">mapIds</span>(org.Hs.eg.db,</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">keys=</span> <span class="fu">row.names</span>(res),</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">keytype=</span> <span class="st">"ENSEMBL"</span>,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">column=</span> <span class="st">"GENENAME"</span>,</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">multiVals=</span><span class="st">"first"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): condition hoxa1_kd vs control_sirna 
Wald test p-value: condition hoxa1 kd vs control sirna 
DataFrame with 10 rows and 9 columns
                 baseMean log2FoldChange     lfcSE       stat      pvalue
                &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000279457   29.9136      0.1803039 0.3121566   0.577607 5.63529e-01
ENSG00000187634  183.2296      0.4258966 0.1355303   3.142446 1.67543e-03
ENSG00000188976 1651.1881     -0.6927118 0.0549876 -12.597612 2.17635e-36
ENSG00000187961  209.6379      0.7299597 0.1277613   5.713463 1.10700e-08
ENSG00000187583   47.2551      0.0392549 0.2606192   0.150622 8.80274e-01
ENSG00000187642   11.9798      0.5395082 0.5001355   1.078724 2.80711e-01
ENSG00000188290  108.9221      2.0562855 0.1910714  10.761870 5.21018e-27
ENSG00000187608  350.7169      0.2570251 0.0999769   2.570845 1.01451e-02
ENSG00000188157 9128.4394      0.3899096 0.0482214   8.085827 6.17439e-16
ENSG00000131591  156.4791      0.1968918 0.1406800   1.399572 1.61641e-01
                       padj      symbol      entrez               genename
                  &lt;numeric&gt; &lt;character&gt; &lt;character&gt;            &lt;character&gt;
ENSG00000279457 6.47026e-01          NA          NA                     NA
ENSG00000187634 3.34029e-03      SAMD11      148398 sterile alpha motif ..
ENSG00000188976 2.35970e-35       NOC2L       26155 NOC2 like nucleolar ..
ENSG00000187961 3.69612e-08      KLHL17      339451 kelch like family me..
ENSG00000187583 9.10931e-01     PLEKHN1       84069 pleckstrin homology ..
ENSG00000187642 3.61174e-01       PERM1       84808 PPARGC1 and ESRR ind..
ENSG00000188290 4.17884e-26        HES4       57801 hes family bHLH tran..
ENSG00000187608 1.79950e-02       ISG15        9636 ISG15 ubiquitin like..
ENSG00000188157 3.15902e-15        AGRN      375790                  agrin
ENSG00000131591 2.23894e-01    C1orf159       54991 chromosome 1 open re..</code></pre>
</div>
</div>
<p>Q. Finally for this section let’s reorder these results by adjusted p-value and save them to a CSV file in your current project directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">order</span>(res<span class="sc">$</span>padj)</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="co">#View(res[ord,])</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res[ord,])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): condition hoxa1_kd vs control_sirna 
Wald test p-value: condition hoxa1 kd vs control sirna 
DataFrame with 6 rows and 9 columns
                 baseMean log2FoldChange     lfcSE      stat    pvalue
                &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSG00000117519   4483.63       -2.42268 0.0612493  -39.5545         0
ENSG00000183508   2053.88        3.20186 0.0728530   43.9496         0
ENSG00000159176   5692.46       -2.31371 0.0589160  -39.2714         0
ENSG00000116016   4423.95       -1.88800 0.0442896  -42.6285         0
ENSG00000164251   2348.77        3.34448 0.0695586   48.0815         0
ENSG00000124766   2576.65        2.39226 0.0623261   38.3829         0
                     padj      symbol      entrez               genename
                &lt;numeric&gt; &lt;character&gt; &lt;character&gt;            &lt;character&gt;
ENSG00000117519         0        CNN3        1266             calponin 3
ENSG00000183508         0      TENT5C       54855 terminal nucleotidyl..
ENSG00000159176         0       CSRP1        1465 cysteine and glycine..
ENSG00000116016         0       EPAS1        2034 endothelial PAS doma..
ENSG00000164251         0       F2RL1        2150 F2R like trypsin rec..
ENSG00000124766         0        SOX4        6659 SRY-box transcriptio..</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(res[ord,], <span class="st">"deseq_results.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install(c("pathview", "gage", "gageData"))</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pathview)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>##############################################################################
Pathview is an open source software package distributed under GNU General
Public License version 3 (GPLv3). Details of GPLv3 is available at
http://www.gnu.org/licenses/gpl-3.0.html. Particullary, users are required to
formally cite the original Pathview paper (not just mention it) in publications
or products. For details, do citation("pathview") within R.

The pathview downloads and uses KEGG data. Non-academic uses may require a KEGG
license agreement (details at http://www.kegg.jp/kegg/legal.html).
##############################################################################</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gage)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gageData)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(kegg.sets.hs)</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sigmet.idx.hs)</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(kegg.sets.hs, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`hsa00232 Caffeine metabolism`
[1] "10"   "1544" "1548" "1549" "1553" "7498" "9"   

$`hsa00983 Drug metabolism - other enzymes`
 [1] "10"     "1066"   "10720"  "10941"  "151531" "1548"   "1549"   "1551"  
 [9] "1553"   "1576"   "1577"   "1806"   "1807"   "1890"   "221223" "2990"  
[17] "3251"   "3614"   "3615"   "3704"   "51733"  "54490"  "54575"  "54576" 
[25] "54577"  "54578"  "54579"  "54600"  "54657"  "54658"  "54659"  "54963" 
[33] "574537" "64816"  "7083"   "7084"   "7172"   "7363"   "7364"   "7365"  
[41] "7366"   "7367"   "7371"   "7372"   "7378"   "7498"   "79799"  "83549" 
[49] "8824"   "8833"   "9"      "978"   

$`hsa01100 Metabolic pathways`
   [1] "10"        "100"       "10007"     "100137049" "10020"     "10026"    
   [7] "100510686" "10063"     "10157"     "10170"     "10195"     "10201"    
  [13] "10229"     "10312"     "10317"     "10327"     "10331"     "1036"     
  [19] "10380"     "10390"     "1040"      "10400"     "10402"     "10423"    
  [25] "10449"     "10476"     "10554"     "10555"     "10558"     "1056"     
  [31] "10588"     "10606"     "10621"     "10622"     "10623"     "10632"    
  [37] "10654"     "1066"      "10678"     "10682"     "10690"     "10714"    
  [43] "10720"     "10768"     "10797"     "10826"     "10841"     "10855"    
  [49] "10873"     "10901"     "10905"     "10941"     "10975"     "10993"    
  [55] "10998"     "11019"     "11041"     "1109"      "11112"     "11128"    
  [61] "1119"      "1120"      "11226"     "11227"     "11232"     "112483"   
  [67] "11253"     "11282"     "11285"     "113026"    "11320"     "11343"    
  [73] "113451"    "113612"    "114805"    "1152"      "1158"      "1159"     
  [79] "1160"      "116285"    "117248"    "119548"    "120227"    "121278"   
  [85] "122481"    "122622"    "123099"    "123745"    "123876"    "124"      
  [91] "124454"    "124975"    "125"       "125061"    "125965"    "125981"   
  [97] "126"       "126328"    "126792"    "127"       "127124"    "128"      
 [103] "128869"    "129607"    "129642"    "130"       "130013"    "131"      
 [109] "1312"      "131669"    "132"       "132158"    "1327"      "132789"   
 [115] "1329"      "1337"      "1339"      "1340"      "134147"    "1345"     
 [121] "1349"      "1350"      "1351"      "135152"    "1352"      "1353"     
 [127] "1355"      "1371"      "1373"      "137964"    "138050"    "138429"   
 [133] "139596"    "140838"    "1431"      "144193"    "144245"    "145226"   
 [139] "146664"    "1491"      "15"        "1503"      "150763"    "151056"   
 [145] "151531"    "1537"      "154141"    "1543"      "1544"      "1548"     
 [151] "1549"      "155066"    "1551"      "1553"      "1555"      "1557"     
 [157] "1558"      "1559"      "1562"      "1571"      "1573"      "157506"   
 [163] "1576"      "1577"      "1579"      "158"       "1581"      "1582"     
 [169] "1583"      "1584"      "1585"      "1586"      "1588"      "1589"     
 [175] "159"       "1593"      "1594"      "1595"      "160287"    "1603"     
 [181] "1606"      "1607"      "1608"      "160851"    "1609"      "1610"     
 [187] "1621"      "162417"    "162466"    "1629"      "1633"      "1635"     
 [193] "1638"      "1644"      "1650"      "166929"    "168391"    "169355"   
 [199] "170712"    "171568"    "1716"      "1717"      "1718"      "1719"     
 [205] "1723"      "1737"      "1738"      "1743"      "1757"      "178"      
 [211] "1786"      "1787"      "1788"      "1789"      "1798"      "18"       
 [217] "1806"      "1807"      "1841"      "1854"      "189"       "1890"     
 [223] "1892"      "191"       "192134"    "1962"      "197258"    "199857"   
 [229] "201595"    "2023"      "2026"      "2027"      "203"       "204"      
 [235] "205"       "2053"      "2058"      "210"       "211"       "212"      
 [241] "2131"      "2132"      "2134"      "2135"      "2137"      "216"      
 [247] "217"       "218"       "2180"      "2181"      "2182"      "2184"     
 [253] "219"       "2194"      "220"       "2203"      "221"       "221223"   
 [259] "221823"    "222"       "2222"      "2224"      "223"       "2235"     
 [265] "224"       "226"       "2271"      "22845"     "22856"     "229"      
 [271] "22928"     "22929"     "22934"     "22978"     "230"       "23057"    
 [277] "231"       "23193"     "23236"     "23305"     "23382"     "23396"    
 [283] "23417"     "23475"     "23483"     "23498"     "23530"     "23545"    
 [289] "23553"     "23556"     "2356"      "23600"     "23649"     "23761"    
 [295] "239"       "240"       "242"       "245972"    "245973"    "246"      
 [301] "246721"    "247"       "248"       "249"       "250"       "251"      
 [307] "2523"      "2524"      "2525"      "2526"      "2527"      "2528"     
 [313] "2529"      "2530"      "2531"      "253558"    "2538"      "2539"     
 [319] "254531"    "2548"      "256435"    "2571"      "2572"      "25796"    
 [325] "2581"      "2582"      "2583"      "25834"     "2584"      "2585"     
 [331] "2588"      "25885"     "2589"      "2590"      "25902"     "2591"     
 [337] "2592"      "259230"    "2593"      "259307"    "2595"      "2597"     
 [343] "26007"     "26035"     "2618"      "262"       "26227"     "26229"    
 [349] "26275"     "26279"     "2628"      "26289"     "2629"      "26290"    
 [355] "26301"     "2632"      "26330"     "2639"      "2643"      "2645"     
 [361] "2650"      "2651"      "2673"      "2678"      "2683"      "2686"     
 [367] "2687"      "270"       "27010"     "27034"     "27087"     "27089"    
 [373] "27090"     "271"       "2710"      "2712"      "27124"     "27165"    
 [379] "272"       "2720"      "27235"     "2729"      "2730"      "27306"    
 [385] "2731"      "27349"     "27430"     "2744"      "2746"      "2747"     
 [391] "275"       "2752"      "276"       "2762"      "277"       "278"      
 [397] "279"       "2799"      "28"        "280"       "2805"      "2806"     
 [403] "2821"      "283208"    "283871"    "284098"    "284541"    "2875"     
 [409] "290"       "29071"     "2937"      "2954"      "29796"     "2987"     
 [415] "29880"     "2990"      "29906"     "29920"     "29922"     "29925"    
 [421] "29926"     "29929"     "29947"     "29958"     "29968"     "30"       
 [427] "3028"      "3030"      "3032"      "3033"      "3034"      "3067"     
 [433] "3073"      "3074"      "3081"      "30814"     "30815"     "30833"    
 [439] "30834"     "3098"      "3099"      "31"        "3101"      "314"      
 [445] "3141"      "3145"      "3155"      "3156"      "3157"      "3158"     
 [451] "316"       "317749"    "32"        "3242"      "3251"      "326625"   
 [457] "3283"      "3284"      "3290"      "3291"      "3292"      "3293"     
 [463] "3294"      "3295"      "33"        "3340"      "3373"      "337876"   
 [469] "339221"    "34"        "340485"    "341392"    "3417"      "3418"     
 [475] "3419"      "341947"    "3420"      "3421"      "3422"      "3423"     
 [481] "3425"      "348158"    "349565"    "35"        "353"       "36"       
 [487] "3612"      "3613"      "3614"      "3615"      "3620"      "3628"     
 [493] "3631"      "3632"      "3633"      "3636"      "37"        "3703"     
 [499] "3704"      "3705"      "3706"      "3707"      "3712"      "374291"   
 [505] "374378"    "3795"      "38"        "383"       "384"       "387787"   
 [511] "39"        "3906"      "391013"    "3938"      "3939"      "3945"     
 [517] "3948"      "3990"      "4047"      "4048"      "4051"      "4056"     
 [523] "411"       "4121"      "4122"      "4124"      "4128"      "4129"     
 [529] "4143"      "4144"      "4190"      "4191"      "4199"      "4245"     
 [535] "4247"      "4248"      "4249"      "427"       "4329"      "435"      
 [541] "4351"      "4357"      "438"       "440"       "440138"    "440567"   
 [547] "441024"    "441531"    "442117"    "445"       "4507"      "4508"     
 [553] "4509"      "4512"      "4513"      "4514"      "4519"      "4522"     
 [559] "4524"      "4535"      "4536"      "4537"      "4538"      "4539"     
 [565] "4540"      "4541"      "4548"      "4594"      "4597"      "4598"     
 [571] "4669"      "4694"      "4695"      "4696"      "4697"      "4698"     
 [577] "47"        "4700"      "4701"      "4702"      "4704"      "4705"     
 [583] "4706"      "4707"      "4708"      "4709"      "471"       "4710"     
 [589] "4711"      "4712"      "4713"      "4714"      "4715"      "4716"     
 [595] "4717"      "4718"      "4719"      "4720"      "4722"      "4723"     
 [601] "4724"      "4725"      "4726"      "4728"      "4729"      "4731"     
 [607] "48"        "4830"      "4831"      "4832"      "4833"      "4837"     
 [613] "4842"      "4843"      "4846"      "4860"      "4907"      "493911"   
 [619] "4942"      "4952"      "4953"      "4967"      "498"       "50"       
 [625] "5009"      "501"       "5033"      "5048"      "50484"     "50487"    
 [631] "5049"      "5050"      "5051"      "5053"      "506"       "50614"    
 [637] "50617"     "50700"     "50814"     "509"       "5091"      "5095"     
 [643] "5096"      "51"        "51004"     "5105"      "51056"     "5106"     
 [649] "51074"     "51082"     "51084"     "51102"     "51109"     "51144"    
 [655] "51166"     "51179"     "51181"     "51196"     "51227"     "51251"    
 [661] "51268"     "513"       "5130"      "51301"     "51380"     "51382"    
 [667] "514"       "51477"     "51478"     "515"       "51540"     "516"      
 [673] "5160"      "51601"     "51604"     "51606"     "5161"      "5162"     
 [679] "5167"      "5169"      "517"       "51703"     "51727"     "51728"    
 [685] "51733"     "51763"     "518"       "51805"     "51809"     "5198"     
 [691] "521"       "5211"      "5213"      "5214"      "522"       "5223"     
 [697] "5224"      "5226"      "523"       "5230"      "5232"      "5236"     
 [703] "525"       "526"       "527"       "5277"      "5279"      "528"      
 [709] "5281"      "5283"      "5286"      "5287"      "5288"      "5289"     
 [715] "529"       "5297"      "5298"      "5313"      "5315"      "5319"     
 [721] "5320"      "5321"      "5322"      "533"       "5330"      "5331"     
 [727] "5332"      "5333"      "5335"      "53354"     "5336"      "5337"     
 [733] "5338"      "534"       "535"       "53630"     "537"       "5372"     
 [739] "5373"      "539"       "53947"     "5406"      "5407"      "5408"     
 [745] "5409"      "54107"     "54187"     "5422"      "5424"      "5425"     
 [751] "5426"      "5427"      "5428"      "5430"      "5431"      "5432"     
 [757] "5433"      "5434"      "54344"     "5435"      "5436"      "54363"    
 [763] "5437"      "5438"      "5439"      "5440"      "5441"      "5444"     
 [769] "5445"      "5446"      "54480"     "54490"     "54575"     "54576"    
 [775] "54577"     "54578"     "54579"     "54600"     "54657"     "54658"    
 [781] "54659"     "54675"     "5471"      "54802"     "548596"    "548644"   
 [787] "549"       "54947"     "54963"     "54965"     "5498"      "54988"    
 [793] "54995"     "55163"     "55191"     "55224"     "55229"     "55256"    
 [799] "55276"     "55300"     "55301"     "55304"     "55312"     "55361"    
 [805] "5538"      "55454"     "55500"     "55512"     "55568"     "5557"     
 [811] "5558"      "55627"     "55650"     "55703"     "55750"     "55753"    
 [817] "55790"     "55808"     "55821"     "55902"     "55907"     "56052"    
 [823] "5625"      "56267"     "5631"      "5634"      "56474"     "56623"    
 [829] "56624"     "56655"     "56848"     "56894"     "56895"     "56898"    
 [835] "56901"     "56913"     "56922"     "56953"     "56994"     "570"      
 [841] "57016"     "57026"     "57134"     "5723"      "5730"      "5740"     
 [847] "5742"      "5743"      "57452"     "574537"    "57678"     "57804"    
 [853] "57818"     "57834"     "5805"      "5831"      "5832"      "5833"     
 [859] "58510"     "5859"      "586"       "5860"      "587"       "593"      
 [865] "594"       "5980"      "60490"     "60495"     "6120"      "6184"     
 [871] "6185"      "622"       "6240"      "6241"      "6296"      "6303"     
 [877] "6307"      "6309"      "6342"      "635"       "6389"      "6390"     
 [883] "6391"      "63917"     "6392"      "64087"     "64131"     "64132"    
 [889] "64409"     "64425"     "6448"      "64579"     "64600"     "646625"   
 [895] "6470"      "6472"      "6476"      "64768"     "6480"      "64802"    
 [901] "64816"     "6482"      "6483"      "6484"      "6487"      "6489"     
 [907] "64902"     "65220"     "65263"     "654364"    "6609"      "661"      
 [913] "6610"      "6611"      "6652"      "6675"      "6677"      "669"      
 [919] "6697"      "6713"      "6718"      "6723"      "683"       "686"      
 [925] "6888"      "6898"      "6916"      "6999"      "7054"      "7083"     
 [931] "7084"      "7086"      "7108"      "7166"      "7167"      "7173"     
 [937] "7263"      "7264"      "729020"    "7298"      "7299"      "7306"     
 [943] "7357"      "7358"      "7360"      "7363"      "7364"      "7365"     
 [949] "7366"      "7367"      "7368"      "7371"      "7372"      "7378"     
 [955] "7381"      "7384"      "7385"      "7386"      "7388"      "7389"     
 [961] "7390"      "7498"      "7841"      "790"       "79053"     "79087"    
 [967] "7915"      "79178"     "7923"      "79369"     "7941"      "79586"    
 [973] "79611"     "79623"     "79646"     "79695"     "79717"     "79796"    
 [979] "79799"     "79814"     "79868"     "79888"     "7991"      "80025"    
 [985] "80055"     "80142"     "80146"     "80201"     "80270"     "80308"    
 [991] "80339"     "80347"     "8050"      "81490"     "81579"     "81616"    
 [997] "81849"     "81888"     "8277"      "8309"      "8310"      "83440"    
[1003] "83549"     "8372"      "8382"      "8394"      "8395"      "8398"     
[1009] "8399"      "84002"     "84076"     "84172"     "84245"     "84265"    
[1015] "84274"     "84284"     "84532"     "84618"     "84620"     "84647"    
[1021] "84649"     "84693"     "847"       "84701"     "84706"     "84720"    
[1027] "84735"     "84803"     "84812"     "84890"     "84920"     "84992"    
[1033] "8509"      "8513"      "8525"      "8526"      "8527"      "8529"     
[1039] "85365"     "8540"      "85465"     "8560"      "8564"      "8566"     
[1045] "8608"      "8611"      "8612"      "8613"      "8630"      "8639"     
[1051] "8659"      "8681"      "8692"      "8693"      "8694"      "8702"     
[1057] "8703"      "8704"      "8705"      "8706"      "8707"      "8708"     
[1063] "873"       "8733"      "874"       "875"       "8760"      "8789"     
[1069] "8790"      "8801"      "8802"      "8803"      "8813"      "8818"     
[1075] "8821"      "883"       "8833"      "8854"      "8867"      "8869"     
[1081] "8871"      "8877"      "8879"      "8942"      "8972"      "8974"     
[1087] "89869"     "8992"      "9"         "90423"     "9060"      "9061"     
[1093] "9091"      "9114"      "91373"     "9162"      "91734"     "9197"     
[1099] "9245"      "92483"     "9249"      "9296"      "93034"     "93183"    
[1105] "9331"      "9348"      "9374"      "9377"      "9380"      "9388"     
[1111] "94005"     "9453"      "9468"      "9487"      "9488"      "9489"     
[1117] "95"        "9514"      "9517"      "952"       "9533"      "9536"     
[1123] "9550"      "9551"      "9563"      "9588"      "9615"      "978"      
[1129] "9791"      "9942"      "9945"     </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>foldchanges <span class="ot">=</span> res<span class="sc">$</span>log2FoldChange</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(foldchanges) <span class="ot">=</span> res<span class="sc">$</span>entrez</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(foldchanges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       &lt;NA&gt;      148398       26155      339451       84069       84808 
 0.18030389  0.42589661 -0.69271181  0.72995968  0.03925487  0.53950816 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>keggres <span class="ot">=</span> <span class="fu">gage</span>(foldchanges, <span class="at">gsets=</span>kegg.sets.hs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">attributes</span>(keggres)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$names
[1] "greater" "less"    "stats"  </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(keggres<span class="sc">$</span>less)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                                  p.geomean stat.mean
hsa04110 Cell cycle                            3.548176e-06 -4.604234
hsa03030 DNA replication                       3.992330e-05 -4.191094
hsa04114 Oocyte meiosis                        2.332810e-04 -3.564509
hsa05130 Pathogenic Escherichia coli infection 6.474555e-04 -3.326950
hsa03440 Homologous recombination              2.248158e-03 -2.967340
hsa03013 RNA transport                         3.743589e-03 -2.698970
                                                      p.val        q.val
hsa04110 Cell cycle                            3.548176e-06 0.0007202798
hsa03030 DNA replication                       3.992330e-05 0.0040522154
hsa04114 Oocyte meiosis                        2.332810e-04 0.0157853509
hsa05130 Pathogenic Escherichia coli infection 6.474555e-04 0.0328583656
hsa03440 Homologous recombination              2.248158e-03 0.0912752208
hsa03013 RNA transport                         3.743589e-03 0.1266580962
                                               set.size         exp1
hsa04110 Cell cycle                                 118 3.548176e-06
hsa03030 DNA replication                             36 3.992330e-05
hsa04114 Oocyte meiosis                              95 2.332810e-04
hsa05130 Pathogenic Escherichia coli infection       46 6.474555e-04
hsa03440 Homologous recombination                    28 2.248158e-03
hsa03013 RNA transport                              140 3.743589e-03</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pathview</span>(<span class="at">gene.data=</span>foldchanges, <span class="at">pathway.id=</span><span class="st">"hsa04110"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa04110.pathview.png</code></pre>
</div>
</div>
<p><img src="hsa04110.pathview.png" class="img-fluid"></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pathview</span>(<span class="at">gene.data=</span>foldchanges, <span class="at">pathway.id=</span><span class="st">"hsa04110"</span>, <span class="at">kegg.native=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: reconcile groups sharing member nodes!</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] 
[1,] "9"  "300"
[2,] "9"  "306"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa04110.pathview.pdf</code></pre>
</div>
</div>
<p><img src="hsa04110.png" class="img-fluid"></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Focus on top 5 upregulated pathways here for demo purposes only</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>keggrespathways <span class="ot">&lt;-</span> <span class="fu">rownames</span>(keggres<span class="sc">$</span>greater)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the 8 character long IDs part of each string</span></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>keggresids <span class="ot">=</span> <span class="fu">substr</span>(keggrespathways, <span class="at">start=</span><span class="dv">1</span>, <span class="at">stop=</span><span class="dv">8</span>)</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>keggresids</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "hsa04142" "hsa05323" "hsa04060" "hsa04640" "hsa05332"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pathview</span>(<span class="at">gene.data=</span>foldchanges, <span class="at">pathway.id=</span>keggresids, <span class="at">species=</span><span class="st">"hsa"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa04142.pathview.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: some node width is different from others, and hence adjusted!</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa05323.pathview.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa04060.pathview.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa04640.pathview.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa05332.pathview.png</code></pre>
</div>
</div>
<p><img src="hsa04142.pathview.png" class="img-fluid"> <img src="hsa05323.pathview.png" class="img-fluid"> <img src="hsa04060.pathview.png" class="img-fluid"> <img src="hsa04640.pathview.png" class="img-fluid"> <img src="hsa05332.pathview.png" class="img-fluid"></p>
<p>Q. Can you do the same procedure as above to plot the pathview figures for the top 5 down-reguled pathways?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>questionpathways <span class="ot">&lt;-</span> <span class="fu">rownames</span>(keggres<span class="sc">$</span>less)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>questioninput <span class="ot">=</span> <span class="fu">substr</span>(questionpathways, <span class="at">start=</span><span class="dv">1</span>, <span class="at">stop=</span><span class="dv">8</span>)</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>questioninput</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "hsa04110" "hsa03030" "hsa04114" "hsa05130" "hsa03440"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pathview</span>(<span class="at">gene.data=</span> foldchanges, <span class="at">pathway.id=</span> questioninput, <span class="at">species=</span><span class="st">"hsa"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa04110.pathview.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa03030.pathview.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa04114.pathview.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa05130.pathview.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Working in directory C:/Users/louis/Downloads/bimm143/R code/class14/class14</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Info: Writing image file hsa03440.pathview.png</code></pre>
</div>
</div>
<p><img src="hsa04110.pathview.png" class="img-fluid"> <img src="hsa03030.pathview.png" class="img-fluid"> <img src="hsa04114.pathview.png" class="img-fluid"> <img src="hsa05130.pathview.png" class="img-fluid"> <img src="hsa03440.pathview.png" class="img-fluid"></p>
<p>d</p>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>